opentelemetrytracer: Dynatrace sampler: Enable adaptative sampling

samohte opened the pull request:
Commit Message: Dynatrace sampler to fetch configuration from an API

Additional Description: In [#32598](https://github.com/envoyproxy/envoy/pull/32598) a custom Dynatrace sampler was added to Envoy. This sampler should now be extended to fetch its configuration via an HTTP call from a Dyantrace cluster.
Testing: Unit, Integration, Manual
Docs Changes: N/A
Release Notes:Dynatrace sampler to fetch configuration from an API
Platform Specific Features: N/A
[Optional Runtime guard:] N/A
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

wbpcode commented on source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:
```diff
@@ -8,10 +10,98 @@ namespace Extensions {
+  }
+}
+
+std::chrono::milliseconds getTimeout(envoy::config::core::v3::HttpUri& http_uri) {
+  auto timeout =
+      std::chrono::milliseconds(DurationUtil::durationToMilliseconds(http_uri.timeout()));
+  // we want to ensure that we don't have more than one pending request
+  if (timeout > (TIMER_INTERVAL / 2)) {
+    timeout = (TIMER_INTERVAL / 2);
+  }
+  return timeout;
+}
```
does this necessary? Seems you will only re-enable the timer when the `onSucess` or `onFailure` is called. So, that means you will only emit a new request after last request is completed?

samohte replied:
You are right, it is not possible to have more than one panding request. I have removed the adjustment in 535d99b7932c49f082bf5e8e1fda969708de1d32

wbpcode commented on source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:
```diff
@@ -8,10 +10,98 @@ namespace Extensions {
+    ENVOY_LOG(warn, "Failed to get sampling configuration from Dynatrace: {}", response_code);
+  }
+
+  if (json_valid || reEnableTimer(static_cast<Http::Code>(response_code))) {
+    timer_->enableTimer(std::chrono::seconds(TIMER_INTERVAL));
+  }
```
Is it reasonable design not to enable the timer if the body is invalid or the response code is unexpected value?

And I think at least here should have an error log to tell user the timer will never be enabled again.

samohte replied:
Response codes lie 400, 401. 403, 404, ... are caused by a config error (wrong uri, wrong/invalid token). Therefore we thought the re-enabling the timer does not make sense.
I agree, we should log an error if we never enable the timer.

wbpcode reviewed (commented):
Thanks for you contribution. And some comments are added. :)

wbpcode commented:
/lgtm api

wbpcode commented:
/wait

samohte commented:
> Thanks for you contribution. And some comments are added. :)

Thanks for your valuable feedback. I addressed your comments in 535d99b7932c49f082bf5e8e1fda969708de1d32

wbpcode commented on source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:
```diff
@@ -8,10 +10,92 @@ namespace Extensions {
+  if (response_code == enumToInt(Http::Code::OK)) {
+    ENVOY_LOG(debug, "Received sampling configuration from Dynatrace: {}",
+              http_response->bodyAsString());
+    json_valid = sampler_config_.parse(http_response->bodyAsString());
```
Will we access the `sampler_config_` at worker threads? If yes, then seems there maybe a contention.

samohte replied:
yes, `sampler_config` is used in worker threads. But it has only two members:
* one is `const` and never modified during runtime.
* the other one is set during `parse` and read in the worker threads. It is an `std::atomic<uint32_t>`, so it should be ok to access the variable in multiple threads.

wbpcode reviewed (changes requested):
Thanks for you in timely update. Now, LGTM overall except the last comment about the thread safe. And please resolve the conflict. Thanks.

samohte commented:
> Thanks for you in timely update. Now, LGTM overall except the last comment about the thread safe. And please resolve the conflict. Thanks.

I have resolved the conflict and answered your comment. Pls. let me know what you think about.

samohte commented:
/retest
